xmtp_id/src/associations/association_log.rs

```diff
@@ -0,0 +1,278 @@
+use super::entity::{Entity, EntityRole};
+use super::hashes::{generate_xid, sha256_string};
+use super::signature::{Signature, SignatureError, SignatureKind};
+use super::state::{AssociationState, StateError};
+
+use thiserror::Error;
+
+// const ALLOWED_CREATE_ENTITY_ROLES: [EntityRole; 2] = [EntityRole::LegacyKey, EntityRole::Address];
+
+#[derive(Debug, Error, PartialEq)]
+pub enum AssociationError {
+    #[error("Error creating association {0}")]
+    Generic(String),
+    #[error("Multiple create operations detected")]
+    MultipleCreate,
+    #[error("XID not yet created")]
+    NotCreated,
+    #[error("Signature validation failed {0}")]
+    Signature(#[from] SignatureError),
+    #[error("State update failed")]
+    StateError(#[from] StateError),
+    #[error("Missing existing member")]
+    MissingExistingMember,
+    #[error("Legacy key is only allowed to be associated using a legacy signature with nonce 0")]
+    LegacySignatureReuse,
+    #[error("Signature not allowed for role {0:?} {1:?}")]
+    SignatureNotAllowed(EntityRole, SignatureKind),
+    #[error("Replay detected")]
+    Replay,
+}
+
+pub trait LogEntry {
+    fn update_state(
+        &self,
+        existing_state: Option<AssociationState>,
+    ) -> Result<AssociationState, AssociationError>;
+    fn hash(&self) -> String;
+}
+
+pub struct CreateXid {
+    pub nonce: u32,
+    pub account_address: String,
+    pub initial_association: AddAssociation,
```

Comment:
With the batching approach, we don't need this anymore - there will be a signature here right?

Reply:
That's right. That change came in after I wrote this, but I'll get it replaced. Should be straightforward.